promql/parser/prettier.go

```diff
@@ -0,0 +1,543 @@
+			if hasImmediateScalar {
+				result += " " + item.Val + " "
+				hasImmediateScalar = false
+			} else {
+				result += p.pd.newLine() + p.pd.pad(nodeInfo.baseIndent) + item.Val
+				if hasGrouping {
+					result += " "
+					p.pd.isNewLineApplied = false
+				} else {
+					result += p.pd.newLine()
+				}
+			}
+		case WITHOUT, BY, IGNORING, ON:
+			if !p.pd.isPreviousBlank(result) {
+				// Edge-case: sum without() (metric_name)
+				result += " "
+			}
+			result += item.Val
+			if item.Typ == BOOL && hasImmediateScalar {
+				result += " "
+			} else if isNodeType(node, aggregateExpr) {
+				p.pd.expectAggregationLabels = true
+			}
+		case BOOL:
+			result += item.Val
+			if hasImmediateScalar {
+				result += " "
+			} else {
+				result += p.pd.newLine()
+			}
+		case EQLC, OFFSET, GROUP_LEFT, GROUP_RIGHT:
+			result += " " + item.Val + " "
+		case COMMA:
+			result += item.Val
+			if (nodeSplittable || hasMultiArgumentCalls) && commaLabelCondition {
+				result += p.pd.newLine()
+			} else if !p.pd.labelsSplittable {
+				result += " "
+			}
+		case COMMENT:
+			if result[len(result)-1] != ' ' {
+				result += " "
+			}
+			if p.pd.isNewLineApplied {
+				result += p.pd.pad(nodeInfo.previousIndent()) + item.Val
+			} else {
+				result += item.Val
+			}
+			p.pd.isPreviousItemComment = true
+		}
+	}
+	if isNodeType(root.node, binaryExpr) && !root.hasScalar {
+		return result, nil
+	} else if isNodeType(root.node, parenExpr) && isChildOfTypeBinary(root.node) {
+		return result, nil
+	}
+	return strings.TrimSpace(result), nil
+}
+
+// refreshLexItems refreshes the contents of the lex slice and the properties
+// within it. This is expected to be called after sorting the lexItems in the
+// pre-format checks.
+func (p *prettier) refreshLexItems(items []Item) []Item {
+	return p.lexItems(stringifyItems(items))
+}
+
+// lexItems converts the given expression into a slice of Items.
+func (p *prettier) lexItems(expression string) (items []Item) {
+	var (
+		l    = Lex(expression)
+		item Item
+	)
+	for l.NextItem(&item); item.Typ != EOF; l.NextItem(&item) {
+		items = append(items, item)
+	}
+	return
+}
+
+func (p *prettier) parseExpr(expression string) error {
+	expr, err := ParseExpr(expression)
+	if err != nil {
+		return errors.Wrap(err, "parse error")
+	}
+	p.Node = expr
+	return nil
+}
+
+// isMetricNameAtomic returns true if the metric name is singular in nature.
+// This is used during sorting of lex items.
+func isMetricNameNotAtomic(metricName string) bool {
+	// We aim to convert __name__ to an ideal metric_name if the value of that labels is atomic.
+	// Since a non-atomic metric_name will contain alphabets other than a-z and A-Z including _,
+	// anything that violates this ceases the formatting of that particular label item.
+	// If this is not done then the output from the prettier might be an un-parsable expression.
+	if regexp.MustCompile("^[a-z_A-Z]+$").MatchString(metricName) && metricName != "bool" {
```

Comment:
Seems like the mentioned regex is not happy with promql_tests. This might be a regression on our promql_tests. The earlier one passed easily, but if I replace with the one given in the link, it gives me an unparsable error. Strange.